R/Spectra-functions.R

```diff
@@ -153,3 +153,109 @@ applyProcessing <- function(object, f = dataStorage(object),
+#'
+#' @param x [Spectra()]
+#'
+#' @param y [Spectra()]
+#'
+#' @param FUN `function` to compare the (m/z matched) intensities from each
+#'     spectrum in `x` with those in `y`.
+#'
+#' @param MAPFUN `function` to map peaks between the two compared spectra. See
+#'     [joinPeaks()].
+#'
+#' @param tolerance `numeric(1)` allowing to define a constant maximal accepted
+#'     difference between m/z values for peaks to be matched.
+#'
+#' @param ppm `numeric(1)` allowing to define a relative, m/z-dependent,
+#'     maximal accepted difference between m/z values for peaks to be matched.
+#'
+#' @param ... additional parameters passed to `FUN` and `MAPFUN`.
+#'
+#' @return
+#'
+#' `matrix` with number of rows equal to length of `x` and number of columns
+#' equal `length(y)`.
+#'
+#' @importFrom stats cor
+#'
+#' @examples
+#'
+#' library(Spectra)
+#' fl <- system.file("TripleTOF-SWATH/PestMix1_SWATH.mzML", package = "msdata")
+#' sps <- Spectra(fl, source = MsBackendMzR())
+#'
+#' sps <- sps[1:10]
+#' res <- .compare_spectra(sps, sps, ppm = 20)
+#' res <- .compare_spectra(sps, sps, method = "spearman", ppm = 40)
+#'
+#' res <- .compare_spectra(sps[1], sps)
+#' res <- .compare_spectra(sps, sps[1])
+#' @noRd
+.compare_spectra <- function(x, y = NULL, MAPFUN = joinPeaks, tolerance = 0,
+                             ppm = 20, FUN = cor, ...) {
+    x_idx <- seq_along(x)
+    y_idx <- seq_along(y)
+    mat <- matrix(NA_real_, nrow = length(x_idx), ncol = length(y_idx),
+                  dimnames = list(spectraNames(x), spectraNames(y)))
+    ## Might need some tuning - bplapply?
+    for (i in x_idx) {
+        for (j in y_idx) {
+            peak_map <- MAPFUN(peaks(x[i])[[1]], peaks(y[j])[[1]],
+                               tolerance = tolerance, ppm = ppm, ...)
+            mat[i, j] <- FUN(peak_map$x[, 2], peak_map$y[, 2], ...)
+        }
+    }
+    mat
+}
+
+#' @description
+#'
+#' Compare each spectrum in `x` with each other spectrum in `x`. Makes use of
+#' `combn` to avoid calculating combinations twice.
+#'
+#' @inheritParams .compare_spectra
+#'
+#' @importFrom utils combn
+#'
+#' @examples
+#'
+#' res <- .compare_spectra(sps, sps)
+#' res_2 <- .compare_spectra_self(sps)
+#' all.equal(res, res_2)    # comparison x[1], x[2] != x[2], x[1]
+#' all.equal(res[!lower.tri(res)], res_2[!lower.tri(res_2)])
+#'
+#' @noRd
+.compare_spectra_self <- function(x, MAPFUN = joinPeaks, tolerance = 0,
+                                  ppm = 20, FUN = cor, ...) {
+    x_idx <- seq_along(x)
+    cb <- combn(x_idx, 2, function(idx) {
+        peak_map <- MAPFUN(peaks(x[idx[1]])[[1]], peaks(x[idx[2]])[[1]],
+                           tolerance = tolerance, ppm = ppm)
+        FUN(peak_map$x[, 2], peak_map$y[, 2], ...)
+    })
+    mat <- matrix(NA_real_, length(x_idx), length(x_idx),
+                  dimnames = list(spectraNames(x), spectraNames(x)))
+    mat[lower.tri(mat)] <- cb
+    for (i in seq_len(nrow(mat))) {
```

Comment:
You could replace the `for` loop by:
```r
# copy lower.tri (needs t and upper.tri because of colum-wise filling) to upper.tri
mat[upper.tri(mat)] <- t(mat)[upper.tri(mat)]
# the comparison function should return the same results if x and y are the same
# so we can just calculate it once and copy it to the whole diagonal
p <- peaks(x[1L])[[1L]][,2L]
diag(mat) <- FUN(p, p, ...)
```

Reply:
```r
p <- peaks(x[1L])[[1L]][,2L]
diag(mat) <- FUN(p, p, ...)
```
is not working for comparison functions that e.g. count common peaks.
So may keep the `for` loop but take the copying out of it, or use `vapply1d` for the diagonal: `diag(mat) <- vapply1d(...)`.
It's up to you.

Reply:
Nice, thanks. Only, I think I can not avoid the for loop: `p <- peaks(x[1L])[[1L]][, 2L]` would only extract the intensities from the first spectrum. I could extract all with `peaks(x)` but would then need a `lapply` or similar already. Instead of `vapply` I would then have to use `mapply` to support functions that calculate something on two numeric vectors (user could provide their own defined functions). Also we would need to pass additional parameters `...` along. In the end I think the `for` loop is easier to understand and should not be much slower than `mapply`.

So, I would use your suggestion with the `upper.tri` and keep the `for` loop for the diagonal.